# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [187]:
import pandas as pd
import numpy as np

## Input Dataset

In [188]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [189]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [190]:
df.head()

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

In [193]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography',	'Gender',	'Age', 'Tenure',	'Balance',	'NumOfProducts',	'HasCrCard',	'IsActiveMember',	'EstimatedSalary',	'Exited']].copy()
df

Geography  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0       France  Female   42       2       0.00              1          1   
1        Spain  Female   41       1   83807.86              1          0   
2       France  Female   42       8  159660.80              3          1   
3       France  Female   39       1       0.00              2          0   
4        Spain  Female   43       2  125510.82              1          1   
...        ...     ...  ...     ...        ...            ...        ...   
9995    France    Male   39       5       0.00              2          1   
9996    France    Male   35      10   57369.61              1          1   
9997    France  Female   36       7       0.00              1          0   
9998   Germany    Male   42       3   75075.31              2          1   
9999    France  Female   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  
0                  1        101348.88       1  
1                  1        112542.58       0  
2                  0        113931.57       1  
3                  0         93826.63       0  
4                  1         79084.10       0  
...              ...              ...     ...  
9995               0         96270.64       0  
9996               1        101699.77       0  
9997               1         42085.58       1  
9998               0         92888.52       1  
9999               0         38190.78       0  

[10000 rows x 10 columns]

In [194]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Geography        10000 non-null  object 
 1   Gender           10000 non-null  object 
 2   Age              10000 non-null  int64  
 3   Tenure           10000 non-null  int64  
 4   Balance          10000 non-null  float64
 5   NumOfProducts    10000 non-null  int64  
 6   HasCrCard        10000 non-null  int64  
 7   IsActiveMember   10000 non-null  int64  
 8   EstimatedSalary  10000 non-null  float64
 9   Exited           10000 non-null  int64  
dtypes: float64(2), int64(6), object(2)
memory usage: 781.4+ KB


## Preprocessing

In [195]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df, columns=["Geography", "Gender"])
df.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Exited  Geography_France  Geography_Germany  \
0        101348.88       1                 1                  0   
1        112542.58       0                 0                  0   
2        113931.57       1                 1                  0   
3         93826.63       0                 1                  0   
4         79084.10       0                 0                  0   

   Geography_Spain  Gender_Female  Gender_Male  
0                0              1            0  
1                1              1            0  
2                0              1            0  
3                0              1            0  
4                1              1            0

In [198]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df.drop("Exited", axis=1)
y = df["Exited"]

In [199]:
X.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0        101348.88                 1                  0                0   
1        112542.58                 0                  0                1   
2        113931.57                 1                  0                0   
3         93826.63                 1                  0                0   
4         79084.10                 0                  0                1   

   Gender_Female  Gender_Male  
0              1            0  
1              1            0  
2              1            0  
3              1            0  
4              1            0

In [200]:
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [202]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [203]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

## Train-Test Split

In [204]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

Model pertama yang digunakan adalah Logistic Regression. Logistic Regression adalah metode statistik yang digunakan untuk memodelkan hubungan antara variabel independen dan variabel dependen yang bersifat biner. Tujuannya adalah untuk memprediksi probabilitas kejadian suatu peristiwa berdasarkan nilai-nilai variabel independen yang diberikan. Model ini menggunakan fungsi logistik atau sigmoid untuk mengubah keluaran linier menjadi probabilitas yang berkisar antara 0 dan 1. Dengan menggunakan teknik optimasi, Logistic Regression mencari parameter yang paling sesuai dengan data pelatihan untuk menghasilkan model yang dapat digunakan untuk klasifikasi data baru.

In [205]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.linear_model import LogisticRegression
model1 = LogisticRegression()
params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model1,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'C': 0.5, 'tol': 0.1}

In [207]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.97      0.89      1983
           1       0.66      0.20      0.31       517

    accuracy                           0.81      2500
   macro avg       0.74      0.59      0.60      2500
weighted avg       0.79      0.81      0.77      2500


[[1929   54]
 [ 412  105]]

0.8136


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

Model kedua yang digunakan adalah K-Nearest Neighbor (KNN) Classifier. K-Nearest Neighbor (KNN) Classifier adalah algoritma machine learning yang digunakan untuk masalah klasifikasi. Dalam KNN, prediksi dilakukan dengan mencari k tetangga terdekat dari data baru berdasarkan jarak Euclidean atau metrik lainnya. Model ini mengklasifikasikan data baru berdasarkan mayoritas label dari tetangga terdekatnya. Nilai k dapat ditentukan untuk mengatur jumlah tetangga yang digunakan dalam prediksi.  

In [209]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.neighbors import KNeighborsClassifier
knn_model = KNeighborsClassifier(n_neighbors=1)
knn_model.fit(X_train, y_train)
#END CODE


KNeighborsClassifier(n_neighbors=1)

In [211]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = knn_model.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.89      0.87      1983
           1       0.51      0.45      0.48       517

    accuracy                           0.80      2500
   macro avg       0.69      0.67      0.68      2500
weighted avg       0.79      0.80      0.79      2500


[[1760  223]
 [ 284  233]]

0.7972


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

Model ketiga yang digunakan adalah Random Forest Classifier. Random Forest Classifier adalah algoritma machine learning yang digunakan untuk masalah klasifikasi. Model ini terdiri dari beberapa pohon keputusan yang bekerja secara independen. Setiap pohon keputusan dibangun dengan menggunakan subset acak dari data pelatihan dan fitur yang dipilih secara acak. Prediksi dilakukan dengan menggabungkan hasil prediksi dari setiap pohon keputusan melalui voting mayoritas. Random Forest Classifier efektif dalam mengatasi overfitting dan dapat menghasilkan prediksi yang akurat dalam berbagai jenis data.

In [212]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.ensemble import RandomForestClassifier
model3 = RandomForestClassifier(n_estimators=390, max_features='auto', random_state=42)
#END CODE

model3.fit(X_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


RandomForestClassifier(max_features='auto', n_estimators=390, random_state=42)

In [216]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = model3.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.96      0.92      1983
           1       0.78      0.49      0.60       517

    accuracy                           0.87      2500
   macro avg       0.83      0.72      0.76      2500
weighted avg       0.86      0.87      0.85      2500


[[1912   71]
 [ 266  251]]

0.8652


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

Berdasarkan ketiga model, model 3 memiliki performa yang paling baik.

Alasan:

1. Akurasi: Model 3 memiliki akurasi sebesar 0.8652, yang berarti model ini mampu memprediksi dengan benar sekitar 86.52% dari total kasus. Akurasi ini lebih tinggi daripada model 1 dan model 2.

2. Precision: Precision adalah metrik yang penting untuk evaluasi model klasifikasi. Model 3 memiliki precision sebesar 0.88 untuk kelas 0 dan 0.78 untuk kelas 1, yang menunjukkan kemampuan model untuk mengklasifikasikan dengan benar masing-masing kelas. Precision ini lebih tinggi daripada model 1 dan model 2.

3. F1-Score: F1-score adalah ukuran yang menggabungkan precision dan recall. Model 3 memiliki F1-score sebesar 0.92 untuk kelas 0 dan 0.60 untuk kelas 1, yang menunjukkan bahwa model ini memiliki keseimbangan yang baik antara precision dan recall untuk kedua kelas. F1-Score ini lebih tinggi daripada model 1 dan model 2.

4. Confusion Matrix: Confusion matrix juga memberikan informasi penting tentang performa model. Model 3 memiliki jumlah false positive (71) dan false negative (266) yang lebih rendah dibandingkan dengan model lainnya, yang menunjukkan kemampuan model untuk menghindari kesalahan prediksi yang signifikan.

Berdasarkan alasan-alasan di atas, model 3 adalah model terbaik karena memiliki performa yang lebih baik dalam hal akurasi, precision, F1-score, dan confusion matrix.